File: python/problem_87.py
from typing import List


def validate_rules(rules: List[str]) -> bool:
    x_lo: dict = {}
    x_hi: dict = {}
    y_lo: dict = {}
    y_hi: dict = {}

    def mx(d, k, v): d[k] = max(d.get(k, float('-inf')), v)
    def mn(d, k, v): d[k] = min(d.get(k, float('inf')),  v)

    for rule in rules:
        parts = rule.split()
        if len(parts) != 3:
            continue
        a, direction, b = parts
        eps = 1.0
        if direction == 'N':   mx(y_lo, a, y_lo.get(b, 0) + eps)
        elif direction == 'S': mn(y_hi, a, y_hi.get(b, 0) - eps)
        elif direction == 'E': mx(x_lo, a, x_lo.get(b, 0) + eps)
        elif direction == 'W': mn(x_hi, a, x_hi.get(b, 0) - eps)
        elif direction == 'NE':
            mx(y_lo, a, y_lo.get(b, 0) + eps)
            mx(x_lo, a, x_lo.get(b, 0) + eps)
        elif direction == 'NW':
            mx(y_lo, a, y_lo.get(b, 0) + eps)
            mn(x_hi, a, x_hi.get(b, 0) - eps)
        elif direction == 'SE':
            mn(y_hi, a, y_hi.get(b, 0) - eps)
            mx(x_lo, a, x_lo.get(b, 0) + eps)
        elif direction == 'SW':
            mn(y_hi, a, y_hi.get(b, 0) - eps)
            mn(x_hi, a, x_hi.get(b, 0) - eps)
        for p in set(list(x_lo.keys()) + list(x_hi.keys())):
            if x_lo.get(p, float('-inf')) >= x_hi.get(p, float('inf')):
                return False
        for p in set(list(y_lo.keys()) + list(y_hi.keys())):
            if y_lo.get(p, float('-inf')) >= y_hi.get(p, float('inf')):
                return False
    return True


def main():
    print("Problem 87:")
    assert not validate_rules(["A N B", "B NE C", "C N A"])
    assert     validate_rules(["A NW B", "A N B"])
    print("All Tests Passed")

File: rust/problem_87.rs
use std::collections::HashMap;

pub fn validate_rules(rules: &[&str]) -> bool {
    let mut x_lo: HashMap<&str,f64> = HashMap::new();
    let mut x_hi: HashMap<&str,f64> = HashMap::new();
    let mut y_lo: HashMap<&str,f64> = HashMap::new();
    let mut y_hi: HashMap<&str,f64> = HashMap::new();
    let mx = |m: &mut HashMap<&str,f64>, k: &'static str, v: f64| { let e = m.entry(k).or_insert(f64::NEG_INFINITY); if v > *e { *e = v; } };
    let mn = |m: &mut HashMap<&str,f64>, k: &'static str, v: f64| { let e = m.entry(k).or_insert(f64::INFINITY); if v < *e { *e = v; } };
    let check = |xl: &HashMap<&str,f64>, xh: &HashMap<&str,f64>, yl: &HashMap<&str,f64>, yh: &HashMap<&str,f64>| -> bool {
        for (k, &lo) in xl { if lo >= *xh.get(k).unwrap_or(&f64::INFINITY) { return false; } }
        for (k, &lo) in yl { if lo >= *yh.get(k).unwrap_or(&f64::INFINITY) { return false; } }
        true
    };
    for rule in rules {
        let parts: Vec<&str> = rule.split_whitespace().collect();
        if parts.len() != 3 { continue; }
        let (a, dir, b) = (parts[0], parts[1], parts[2]);
        let eps = 1.0f64;
        let yb = *y_lo.get(b).unwrap_or(&0.0); let yt = *y_hi.get(b).unwrap_or(&0.0);
        let xb = *x_lo.get(b).unwrap_or(&0.0); let xt = *x_hi.get(b).unwrap_or(&0.0);
        match dir {
            "N"  => { let v=yb+eps; *y_lo.entry(a).or_insert(f64::NEG_INFINITY) = v.max(*y_lo.get(a).unwrap_or(&f64::NEG_INFINITY)); }
            "S"  => { let v=yt-eps; *y_hi.entry(a).or_insert(f64::INFINITY)     = v.min(*y_hi.get(a).unwrap_or(&f64::INFINITY)); }
            "E"  => { let v=xb+eps; *x_lo.entry(a).or_insert(f64::NEG_INFINITY) = v.max(*x_lo.get(a).unwrap_or(&f64::NEG_INFINITY)); }
            "W"  => { let v=xt-eps; *x_hi.entry(a).or_insert(f64::INFINITY)     = v.min(*x_hi.get(a).unwrap_or(&f64::INFINITY)); }
            "NE" => {
                let vy=yb+eps; *y_lo.entry(a).or_insert(f64::NEG_INFINITY) = vy.max(*y_lo.get(a).unwrap_or(&f64::NEG_INFINITY));
                let vx=xb+eps; *x_lo.entry(a).or_insert(f64::NEG_INFINITY) = vx.max(*x_lo.get(a).unwrap_or(&f64::NEG_INFINITY));
            }
            "NW" => {
                let vy=yb+eps; *y_lo.entry(a).or_insert(f64::NEG_INFINITY) = vy.max(*y_lo.get(a).unwrap_or(&f64::NEG_INFINITY));
                let vx=xt-eps; *x_hi.entry(a).or_insert(f64::INFINITY)     = vx.min(*x_hi.get(a).unwrap_or(&f64::INFINITY));
            }
            "SE" => {
                let vy=yt-eps; *y_hi.entry(a).or_insert(f64::INFINITY)     = vy.min(*y_hi.get(a).unwrap_or(&f64::INFINITY));
                let vx=xb+eps; *x_lo.entry(a).or_insert(f64::NEG_INFINITY) = vx.max(*x_lo.get(a).unwrap_or(&f64::NEG_INFINITY));
            }
            "SW" => {
                let vy=yt-eps; *y_hi.entry(a).or_insert(f64::INFINITY)     = vy.min(*y_hi.get(a).unwrap_or(&f64::INFINITY));
                let vx=xt-eps; *x_hi.entry(a).or_insert(f64::INFINITY)     = vx.min(*x_hi.get(a).unwrap_or(&f64::INFINITY));
            }
            _ => {}
        }
        if !check(&x_lo, &x_hi, &y_lo, &y_hi) { return false; }
    }
    true
}

pub fn problem87() {
    println!("Problem 87:");
    assert!( validate_rules(&["A N B","B NE C","C N A"]) == false);
    assert!( validate_rules(&["A NW B","A N B"]));
    println!("All Tests Passed");
}
